internal/tar_interpreter_new.go

```diff
@@ -11,69 +11,62 @@ import (
+	fileUnwrapper := getFileUnwrapper(tarInterpreter, header, targetPath)
+	if localFileInfo, _ := getLocalFileInfo(targetPath); localFileInfo != nil {
+		return unwrapToExistFile(fileReader, header, targetPath, fileUnwrapper)
+	}
+	return unwrapToNewFile(fileReader, header, targetPath, fileUnwrapper)
+}
+
+// get file unwrapper for file depending on backup type
+func getFileUnwrapper(tarInterpreter *FileTarInterpreter, header *tar.Header, targetPath string) IBackupFileUnwrapper {
+	fileDescription, haveFileDescription := tarInterpreter.Sentinel.Files[header.Name]
 	isIncremented := haveFileDescription && fileDescription.IsIncremented
+	var isPageFile bool
+	if localFileInfo, _ := getLocalFileInfo(targetPath); localFileInfo != nil {
+		isPageFile = isPagedFile(localFileInfo, targetPath)
+	}
+	options := &BackupFileOptions{isIncremented: isIncremented, isPageFile: isPageFile}
+
 	// todo: clearer catchup backup detection logic
 	isCatchup := tarInterpreter.createNewIncrementalFiles
-
-	if localFileInfo, _ := getLocalFileInfo(targetPath); localFileInfo != nil {
-		isPageFile := isPagedFile(localFileInfo, targetPath)
-		return unwrapToExistFile(fileReader, fileInfo, targetPath, isPageFile, isIncremented, isCatchup)
+	if isCatchup {
+		return NewFileUnwrapper(CatchupBackupFileUnwrapper, options)
 	}
-	return unwrapToNewFile(fileReader, fileInfo, targetPath, isIncremented)
+	return NewFileUnwrapper(DefaultBackupFileUnwrapper, options)
 }
 
 // unwrap the file from tar to existing local file
-func unwrapToExistFile(fileReader io.Reader, fileInfo *tar.Header, targetPath string,
-	isPageFile, isIncremented, isCatchup bool) error {
+func unwrapToExistFile(fileReader io.Reader, header *tar.Header, targetPath string, unwrapper IBackupFileUnwrapper) error {
 	localFile, err := os.OpenFile(targetPath, os.O_RDWR, 0666)
 	if err != nil {
 		return err
 	}
 	defer utility.LoggedClose(localFile, "")
-	if isIncremented {
-		err := WritePagesFromIncrement(fileReader, localFile, isCatchup)
-		return errors.Wrapf(err, "Interpret: failed to write increment to file '%s'", targetPath)
-	}
-	if isCatchup {
-		err := clearLocalFile(localFile)
-		if err != nil {
-			return err
-		}
-		return writeLocalFile(fileReader, fileInfo, localFile)
-	}
-	if isPageFile {
-		err := RestoreMissingPages(fileReader, localFile)
-		return errors.Wrapf(err, "Interpret: failed to restore pages for file '%s'", targetPath)
-	}
-	// skip the non-page file because newer version is already on the disk
-	return nil
+	return unwrapper.UnwrapExistingFile(fileReader, header, localFile)
 }
 
 // unwrap file from tar to new local file
-func unwrapToNewFile(fileReader io.Reader, fileInfo *tar.Header, targetPath string, isIncremented bool) error {
-	localFile, err := createLocalFile(targetPath, fileInfo)
+func unwrapToNewFile(fileReader io.Reader, header *tar.Header, targetPath string, unwrapper IBackupFileUnwrapper) error {
```

Comment:
It's used only in one place. And as I got it doesn't contain specific logic except for open file. Now it looks like a method of IBackupFileUnwrapper. But actually it isn't.

Reply:
So, I assume there is two ways to solve this:
1. Move this file creating/opening logic to IBackupFileUnwrapper 
2. Move this back to the place it is called and just get rid of the separate function

Which will be better?

Reply:
I vote for second

Reply:
fixed in 4a91a430a19ebdcae18eebccf9a061de16a85156